control-plane/p4RuntimeSerializer.cpp

```diff
@@ -766,13 +783,22 @@ class P4RuntimeAnalyzer {
             addDocumentation(param, actionParam->to<IR::IAnnotated>());
 
             auto paramType = typeMap->getType(actionParam, true);
-            if (!paramType->is<IR::Type_Bits>() && !paramType->is<IR::Type_Boolean>()) {
-                ::error("Action parameter %1% has a type which is not bit<> or int<> or bool",
-                        actionParam);
+            if (!paramType->is<IR::Type_Bits>() && !paramType->is<IR::Type_Boolean>()
+                && !paramType->is<IR::Type_Newtype>()) {
+                ::error("Action parameter %1% has a type which is not "
+                        "bit<> or int<> or bool or New type", actionParam);
                 continue;
             }
             if (paramType->is<IR::Type_Boolean>()) {
                 param->set_bitwidth(1);
+            } else if (paramType->is<IR::Type_Newtype>()) {
```

Comment:
same comment as above, I think `width_bits` should work for custom types as well. If this is not the case, `width_bits` should be "fixed" or an equivalent recursive helper should be written to be used by this code.

Reply:
I have updated the PR with new code changes.  Both places in code use the same new function that now works for deeply nested type and typdef mixed together.